### DISCUSSION TITLE
Calculate DeepSleepTime wake time

### DISCUSSION THREAD

**Discussion by smhc** [arendst/Tasmota#20042] [Q&A] [answered]
I think I understand how the DeepSleepTime works now, but it was quite difficult to decipher as the document suggests there is a max sleep time of 86400 and that value wakes at 00:00, somewhat implying that you can only sleep until midnight as a maximum. 

But this documentation is incorrect as you can specify up to 10 years of sleep time. Also, depending on what you specify and when you perform the sleep - specifying a smaller sleep time can actually result in waking later.

It also hints that there's no way to 'shift' the wake up time and the difficulties of targeting a particular time - but really that is referring to the next wake up time in a recurring set. If you don't care about the recurring period you can certainly target a given *first* wake up time if you happen on just the right value. You can always ignore the recurring period and recalculate the next wake up time yourself.

This is combined with the fact the documentation doesn't mention the 5% fudge factor, which I believe is a bit buggy/excessive as 86400 actually wakes at 01:12 the next day. I think this is quite excessive buffer amount.

I've listed below some relatively random sleep times and the local time that they were performed. Something like this might be useful in the doco.

DeepSleepTime | Hours | 5% mins | Current local time | First WakeTime
-- | -- | -- | -- | --
16200 | 4.5 | 13.5 | 2023-11-17 12:28:17 | 2023-11-17 16:43:30
100800 | 28 | 84 | 2023-11-17 12:29:53 | 2023-11-18 09:24:00
90000 | 25 | 75 | 2023-11-17 12:39:45 | 2023-11-18 05:15:00
86520 | 24.03333333 | 72.1 | 2023-11-17 12:42:45 | 2023-11-18 08:16:06
3600 | 1 | 3 | 2023-11-17 12:43:39 | 2023-11-17 13:03:00
86400 | 24 | 72 | 2023-11-17 12:51:14 | 2023-11-18 01:12:00
82800 | 23 | 69 | 2023-11-17 12:53:33 | 2023-11-18 10:09:00
12600 | 3.5 | 10.5 | 2023-11-17 12:56:32 | 2023-11-17 14:40:30
187200 | 52 | 156 | 2023-11-17 13:00:39 | 2023-11-18 22:36:00
374400 | 104 | 312 | 2023-11-17 13:02:20 | 2023-11-21 05:12:00
16200 | 4.5 | 13.5 | 2023-11-17 13:03:41 | 2023-11-17 16:43:30
3600 | 1 | 3 | 2023-11-17 13:09:01 | 2023-11-17 14:03:00
90000 | 25 | 75 | 2023-11-17 13:14:35 | 2023-11-18 05:15:00

It would be great if Tasmota had a "DeepSleepFor" or "DeepSleepUntil" command that slept for x seconds or slept until a given local time. But anyway, I will work with this current command which is really more like a "DeepSleepSchedule".

My question is whether anyone has a formula for calculating this "first wake time" given a current time and deep sleep time? Or even better, calculate the largest deep sleep time given a current time and target wake time? I would like to specify a target wake time for just a once off wake - I can then recalculate the next deep sleep amount as required for the next wake (e.g in berry).

I understand it is periods of deep sleep time from the epoch up until it exceeds the current time, then 1 more deep sleep time unit after that. But I cannot think of a way to calculate this.

I believe calculating the ideal deep sleep time might be finding the largest factor of 'floor(timestamp(target_time) + x mins)'. The reason I want to use the largest value is to avoid any interim wakes between now and the target time. I can round the target time up/down 10 minutes or so if that makes finding a suitable factor easier. Any other ideas?

**Comment by smhc** [comment]
I thought that taking the target timestamp and dividing by 8.4 (getting a factor within the valid range, then reducing by the 5%) would give you an appropriate "deep sleep time". But that doesn't seem to work.

I believe the formula is more or less:

waketime = ((curtime / dsleep) + 1) * dsleep) + (dsleep * 0.05)

based on : https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_29_deepsleep.ino#L117
(fixed link: https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_29_deepsleep.ino#L127 )

but I am not maths enough to figure it out :(
Based on chatgpt I think the formula might be something like:

dsleep = (waketime - curtime) / 1.05

but it's not quite right for me.

**Reply by stefanbode** [reply]
The formula is correct because there is a temperature correction that takes into account what the actual sleep time was with the corresponding factor and calculate a new one. Your suggestion of the formula does not work as far as I can see. There where requests to shift the time or make something on top for timezones because the sleeptime is always compared to the UNIX time in seconds since epoch. This makes is impossible to wake up daily at 3am. The wakeup time with SHIFT over time because with every run it get a bit better in prediction how long it has to sleep. Because this is temperature dependent there are limits. The is a graph in the docu that shows the compensation benefits over time. It need some cycles to get the correct one. The table you provided seems wrong to me. The 1.05 has a total different meaning. It is just to see if it was a full cycle or a manual wakeup or just a fraction of the cycle to calculate the correct compensation factor. 

3600 will always try to wakeup exactly at 1:00am, 2:00am and so on.
7200 at 2pm 4pm 6pm and so on

**Reply by smhc** [reply]
The table is based on actually running "DeepSleepTime x" at the given current time and taking note of the "WakeupTime" as reported via the MQTT message. I may have made the odd copy-paste mistake, but for the most part it should reflect the actuals, at least as far as the self-reporting is concerned.

My link to the code was wrong, it should have been: https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_29_deepsleep.ino#L127

and indeed if you look at the wake times they are all increased by 5%, as per the code. (86400 x 5% = 1h 12min)

I am not too concerned about "every run" or the slipping or waking up at 3am daily. I am only concerned about the "first wake". After that point I will cancel the deep sleep and calculate the next wake. Essentially I am abusing the 'deep sleep schedule period' functionality to perform a 'deep sleep wake once at x'.

From what I can see in the code after calling "DeepSleepTime x" the nextwakeup is 0, which causes it to go into this code: https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_29_deepsleep.ino#L127

At which point the formula seems to be : 

waketime = (((floor((1700188097+ 5) / 16200)) + 1) * 16200) + (16200* 0.05)

But I am obviously missing something as this doesn't seem to calculate out correctly. Possibly a time zone issue.

**Comment by stefanbode** [comment]
deepsleep 16200 is a bad choice because this in epoch seconds is every day a different time. Anyhow the time between the wakeups is more or less exactly 16200 seconds. You get the current epoch time here: https://www.epochconverter.com/

e.g. now: 1700206164  --> 1700206164/16200 = 104.950 full cycle -> wakeup at 104.951 * 16200 timestamp

Your time zone: Freitag, 17. November 2023 08:28:38 [GMT+01:00] -> wakeup : Freitag, 17. November 2023 08:30:00 [GMT+01:00]
this is in this case just in two minutes.

**Comment by smhc** [comment]
16200 is a bad choice for waking 2 or more times, but as far as waking once all I care about is the first wake time. After that I will cancel the sleep schedule and issue a new deep sleep command with a new value.

I think your calculation is more or less what I was performing (except I was also taking into account the 5% bump)- but I was not getting the same values as "DeepSleepTime x" was reporting.

**Comment by smhc** [comment]
EDIT: ChatGPT is a terrible math tool. Using a excel I get the correct value

= (( ( floor ((1700224097+ 5) / 16200)) + 1) * 16200) + (16200* 0.05)
= 1700239410
= 2023-11-17 16:43:30

**Reply by smhc** [reply]
And ChatGPT gives the following for solving for dsleep, which isn't very promising. I guess it might actually require some trial and error to solve.



Let's solve for `dsleep` in the given equation:

\[ 1700238700 = \left( \left( \text{floor}\left(\frac{1700224102 + 5}{dsleep}\right) + 1 \right) \cdot dsleep \right) + (dsleep \cdot 0.05) + dsleep \]

First, calculate the value inside the floor function:

\[ \text{floor}\left(\frac{1700224107}{dsleep}\right) = \text{floor}\left(\frac{1700224107}{dsleep}\right) \]

Now, substitute this value back into the expression:

\[ 1700238700 = \left( \left(\text{floor}\left(\frac{1700224107}{dsleep}\right) + 1 \right) \cdot dsleep \right) + (dsleep \cdot 0.05) + dsleep \]

This equation involves a floor function, which makes it challenging to solve algebraically in a general sense. Depending on the specific values and constraints, you might need to use numerical methods or trial-and-error to find a suitable value for `dsleep` that satisfies the equation. If there are additional constraints or context for the problem, that information could be useful in finding a solution.

**Reply by smhc** [reply]
That said that formula still fails for many other rows in the table so it's still not correct.
EDIT - I think it's correct, but needed to use the right tools

**Reply by stefanbode** [reply]
Where did you get the 1.05? Wake-up is sharp in time

  if (timeslip) {
    RtcSettings.nextwakeup += Settings->deepsleep;
    RtcSettings.deepsleep_slip = (RtcSettings.nextwakeup - LocalTime()) * RtcSettings.deepsleep_slip / tmax((Settings->deepsleep - (millis() / 1000)),5);
    // Avoid crazy numbers. Again maximum 10% deviation.
    RtcSettings.deepsleep_slip = tmin(tmax(RtcSettings.deepsleep_slip, 9000), 11000);
  }

The slip is calculated and is neutral at first start. Deepsleep shift if any is because of drift of the internal clock.

**Reply by smhc** [reply]
I had corrected my link to the right line of code: 
https://github.com/arendst/Tasmota/blob/74aa056650e2bfb9a3c4910bad3c14af38a2b51d/tasmota/tasmota_xdrv_driver/xdrv_29_deepsleep.ino#L128C6-L128C58

```C
RtcSettings.nextwakeup += Settings->deepsleep * 0.05;
```

It's not part of the timeslip calculation, but rather baked into the first wake calculation to ensure it is always slightly ahead and never slightly behind. Apparently this was important for synchronisation with other systems: https://github.com/arendst/Tasmota/pull/17016.  As I mentioned I think this is probably excessive, especially for the 86400 case which is supposed to wake at 00:00 but actually ends up being 01:12.

I'm a little concerned that this won't always be applied, as it only gets into this block if nextwakeup is 0, which it might not be due to the code you referred to:

https://github.com/arendst/Tasmota/blob/74aa056650e2bfb9a3c4910bad3c14af38a2b51d/tasmota/tasmota_xdrv_driver/xdrv_29_deepsleep.ino#L116

**Reply by stefanbode** [reply]
The wake-up after 72 minutes is mandatory. Anyhow the device will be up for less than a second and go for another hour. Only when the time has come a complete wake-up with WiFi will occur. There is no way around possible not required. 

What is your problem with the slip? If the wake-up is in a window of 95% to 105% it assumes this is NOT a manual wake-up and calculate the slip to wake-up at the right time the next cycle. The slip is only for compensation. As you see in the documentation this works excellent. 

Can you send a device with deepsleep at any time to a e.g. 3600 deepsleep? Answer is No!
It will try to wake-up at he next full hour no matter when you start the deepsleep.

**Reply by smhc** [reply]
I don't have an issue with the slip, I assume this is required between cycles. It is the very first wake that is getting offset by 5% that is potentially an issue. If you perform a "DeepSleepTime 86400" intending to wake up at midnight each night (as per doco) - it will actually wake up at 01:12 each night instead, an hour and 12 minutes late. Or at least, that is what the first nextwakup calculation and logging suggests will happen. Maybe this reported wakeup time is actually incorrect and the real wakeup time gets corrected during the mandatory "mini"-wakes?

There are two expression blocks in the code - one for the slip, another for the 'first wake'. The 5% bump was added to offset wakes to ensure they are always ahead, to fix some home assistant integration issue. IMHO it probably should have been either capped to ~3 minutes or fixed on the home assistant end. 

> Can you send a device with deepsleep at any time to a e.g. 3600 deepsleep? Answer is No!

Is this in reference to the current implementation, or a limitation of the ESP hardware? Couldn't an end user use that berry code I posted below to perform a sleep for current_time+3600 to effectively achieve this? I am ignoring the interval/cycles - you'd need to have an init rule to calculate a new deepsleep time each wake. (I am doing this for https://github.com/arendst/Tasmota/discussions/19977 and it appears to work ok).

In other words - if we were to expose esp_sleep_enable_timer_wakeup to berry could we conceivably sleep until any time and not have this 'floor based' interval scheduling? I think that would be very useful if so.

**Reply by smhc** [reply]
For reference, my use case is I want to "full wake" twice a day and only twice a day to save power. Ideally at 5:00am and 4:30pm (or possibly sunrise and sunset). I believe the best way to achieve this is with a bit of berry code that calculates the sleep time for the next "DeepSleepTime" command on each wake. I believe the "sleep_until" bit of code I posted below could be used to achieve the "sleep until specific time" side of things.

I have implemented this under https://github.com/arendst/Tasmota/discussions/19977 and it seems to work ok, only waking at the appropriate target time from an arbitrary starting time.

**Reply by stefanbode** [reply]
We checked a few month ago to get this into the timer config. So you have on/off/toggle/rule/deepsleep as an event. This would 100% match your requirements. In this case you can also make use of the sunset. But....
This entry is only 2bit and allows only 4 states and not 5.
For the esp8266 unlikely to change. For esp32 I have to check with arendst.

**Reply by smhc** [reply]
Yeah that would be ideal, and I imagine quite useful for many other applications. Although I guess it's more of a 'wake' setting than a sleep setting. You could put the device to sleep with 'Restart 9' or whatever, and that timer would ideally dictate when it gets configured to wake again.

Not sure if the 2bit limit you are referring to is this: https://github.com/arendst/Tasmota/blob/development/tasmota/include/tasmota_types.h#L333
~~But it looks like the .time member would only need values up to 1440, so plenty of values above that that fit in 11 bits which could be used to indicate other flags maybe?~~ (this is wrong) Anyway, perhaps this discussion would be better on the respective feature issue.

**Reply by stefanbode** [reply]
Yes line 333 is the issue kid. If we can store the wake-up time here it is easy to calculate on demand the deepsleep time based on current time and the timers. The timers in this case are just used as an UI and a bucket to save the information. Calculation will be externalized in the deepsleep driver with a Formular similar to yours.

**Comment by smhc** [comment]
Here is a spreadsheet that takes a given 'current time' and shows the 'first wake time' values for various values of 'deep sleep time'. Ideally I need a way to do the reverse - calculate the deep sleep time given a desired wake time.
https://docs.google.com/spreadsheets/d/1NpH8qJoR9OpKxXcYKQXaIeRDBoOZcjo70IRB5QNVGJM/edit?usp=sharing

As you can see, unless I am mistaken, things get pretty weird and unpredictable with values above 20,000 or so.

**Comment by smhc** [comment, marked as answer]
This is a bit of berry code that will sleep until a specific time:

```berry
import math
import string
def sleep_until(target)
    var curtime = tasmota.rtc()["local"]

    var ds = 60
    for i: 1 .. 10000
        var wake = (((curtime / ds) + 1) * ds ) + (ds * 0.05)
        if (math.abs(wake - target) < 600)
            # print('sleep time: ', ds, i)
            tasmota.cmd(string.format("DeepSleepTime %d", ds))
            return i
        end
        if (i % 1000 == 0)
            tasmota.yield()
        end
        ds += 60
    end
    print('not found..')
end
# Sleep until 12 hours after the current time
sleep_until(tasmota.rtc()["local"] + (3600 * 12))
```

It's not that efficient, and would need some tuning of the search iteration depending on how long you needed to sleep for. But it works ok for < 24hr sleeps.

The sleep time chosen is unlikely to be suitable for the "next" period. You will need to have a rule that immediately disables DeepSleepTime on init, or immediately sets it to the new desired value. If Tasmota ever changes the sleep time algo it will likely break.

Would be nice to have native support for this functionality in Tasmota, but alas this is ok for now.

**Reply by sfromis** [reply]
Long loops of 10000 is strongly not recommended in Berry, as code should always return quickly for Tasmota to be responsive, ideally under 50 ms. Using `tasmota.yield()` does not relinquish control as Tasmota runs on a single thread. It will only keep the watchdog from biting, but not keep Tasmota operational.

You cannot make a good `sleep_until` function in Tasmota+Berry. What you should do is to use `tasmota.set_timer` for when you want to resume, or set up a `cron` timer.

**Reply by smhc** [reply]
Yeah fair enough - although it typically only loops 1000 times or so before it finds a suitable value and early exits. It's just doing a quick maths operation and returning, so hopefully under 50ms. (I timed this and it's generally only 15ms or so to find a value). If there was a way to inverse that equation the loop wouldn't be required at all. But I haven't figured out a way to do that (I'm not sure it's possible). It also end with a deepsleep, so may not be that critical if it blocks tasmota in it's final ms before shutting down.

> You cannot make a good sleep_until function in Tasmota+Berry.

While it's named 'sleep until', keep in mind it isn't really "sleeping", it's just calculating a value to pass to deepsleeptime. You could rework it to use the 'fast_driver' functionality with a counter kept in state if it exceeds 50ms.

**Comment by stefanbode** [comment]
Christmas is comming!! @smhc 

Added a new pull request to use TIMERS for wakeup from deepsleep. The new add on calculates the next wakeup based on the TIMERS. You can use nearly all TIMERS functionality beside the RANDOM change.

Pull request: https://github.com/arendst/Tasmota/pull/20117

How does it work (must be rule1 ther rules 2,3 do not work):
`rule1 Wakeup`
`rule1 on`

Define TIMERS with action RULE

New functionality: On power restart and reset Deepsleepcountdown is 60seconds. after deepsleep wakeup 4 seconds

**Reply by stefanbode** [reply]
I also removed the 5% add on on the initial run and just take care that there are more than 15 seconds before the next wakeup.

**Reply by stefanbode** [reply]
Documentation updated: https://github.com/tasmota/docs/blob/master/docs/DeepSleep.md

**Reply by stefanbode** [reply]
#20165 submitted to add "restart 9" as an option to send the device to deepsleep. Documentation will follow.

**Comment by stefanbode** [comment]
Documentation updated: https://tasmota.github.io/docs/DeepSleep/#deepsleep-wakeup-based-on-timer-events
